File: src/design_patterns/behavioral/template_method/javascript/template_method.js
class AbstractClass {
    templateMethod() {
        this.operation1();
        this.operation2();
        this.operation3();
    }

    operation1() { }

    operation2() { }

    operation3() { }
}

class ConcreteClass extends AbstractClass {
    operation1() {
        console.log("ConcreteClass.operation1() called");
    }

    operation2() {
        console.log("ConcreteClass.operation2() called");
    }
}

class AnotherConcreteClass extends AbstractClass {
    operation1() {
        console.log("AnotherConcreteClass.operation1() called");
    }

    operation3() {
        console.log("AnotherConcreteClass.operation3() called");
    }
}

File: src/design_patterns/behavioral/template_method/c#/template_method.cs
abstract class AbstractClass
{
    public void TemplateMethod()
    {
        Operation1();
        Operation2();
        Operation3();
    }

    protected virtual void Operation1() { }

    protected virtual void Operation2() { }

    protected virtual void Operation3() { }
}

class ConcreteClass : AbstractClass
{
    protected override void Operation1()
    {
        Console.WriteLine("ConcreteClass.Operation1() called");
    }

    protected override void Operation2()
    {
        Console.WriteLine("ConcreteClass.Operation2() called");
    }
}

class AnotherConcreteClass : AbstractClass
{
    protected override void Operation1()
    {
        Console.WriteLine("AnotherConcreteClass.Operation1() called");
    }

    protected override void Operation3()
    {
        Console.WriteLine("AnotherConcreteClass.Operation3() called");
    }
}
